sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -88,13 +88,15 @@ case class AdaptiveSparkPlanExec(
   private def queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
     RemoveRedundantProjects,
     EnsureRequirements,
-    RemoveRedundantSorts,
     DisableUnnecessaryBucketedScan
   ) ++ context.session.sessionState.queryStagePrepRules
 
   // A list of physical optimizer rules to be applied to a new stage before its execution. These
   // optimizations should be stage-independent.
   @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
+    // We should remove redundant sort after reOptimizer since reOptimize will change SMJ to BHJ
+    // which can affect the output ordering
```

Comment:
doesn't `queryStagePreparationRules` run right after reoptimization?

Reply:
`initialPlan` is a special case which run before the reoptimization.

The root cause of this issue is the top level `SortExec` node removed at `initialPlan`, then the `Sort` which also at top level lost. The key code is `currentPhysicalPlan.logicalLink.get`.